trigger/twister2.py

```diff
@@ -0,0 +1,445 @@
+        self.data = ''
+        log.msg('[%s] connectionMade, data: %r' % (self.device, self.data))
+        # self.factory._init_commands(self)
+        # self._send_next()
+
+
+    def connectionLost(self, reason):
+        self.finished.callback(None)
+
+
+        # Don't call _send_next, since we expect to see a prompt, which
+        # will kick off initialization.
+
+    def add_commands(self, commands):
+        self._init_results_map(commands)
+        # self.commands_entered = self.commands_entered + commands
+        self.remaining_commands = commands[len(self.commands_entered):]
+        if len(self.remaining_commands) > 0:
+            self.commanditer = iter(self.remaining_commands)
+        else:
+            self.commands = commands
+            self.commanditer = iter(commands)
+        return True
+
+    def _init_results_map(self, commands):
+        self._results_lock = defer.DeferredLock()
+        self.commands_epoch.append(commands)
+        self._results_lock.acquire()
+        self.results_map[hash_list(commands)] = {tuple(commands): None}
+        self._results_lock.release()
+
+    def set_results_map(self, results, commands):
+        self._results_lock.acquire()
+        self.results_map[hash_list(commands)][tuple(commands)] = copy(results)
+        self._results_lock.release()
+
+    def get_results_map(self, commands):
+        self._results_lock.acquire()
+        rv = self.results_map.get(hash_list(commands))
+        self._results_lock.release()
+        return rv
+
+    def dataReceived(self, bytes):
+        """Do this when we get data."""
+        log.msg('[%s] BYTES: %r' % (self.device, bytes))
+        self.data += bytes
+
+        # See if the prompt matches, and if it doesn't, see if it is waiting
+        # for more input (like a [y/n]) prompt), and continue, otherwise return
+        # None
+        m = self.prompt.search(self.data)
+        if not m:
+            # If the prompt confirms set the index to the matched bytes,
+            if is_awaiting_confirmation(self.data):
+                log.msg('[%s] Got confirmation prompt: %r' % (self.device,
+                                                              self.data))
+                prompt_idx = self.data.find(bytes)
+            else:
+                return None
+        else:
+            # Or just use the matched regex object...
+            prompt_idx = m.start()
+
+        result = self.data[:prompt_idx]
+        # Trim off the echoed-back command.  This should *not* be necessary
+        # since the telnet session is in WONT ECHO.  This is confirmed with
+        # a packet trace, and running self.transport.dont(ECHO) from
+        # connectionMade() returns an AlreadyDisabled error.  What's up?
+        log.msg('[%s] result BEFORE: %r' % (self.device, result))
+        result = result[result.find('\n')+1:]
+        log.msg('[%s] result AFTER: %r' % (self.device, result))
+
+        if self.initialized and result != '':
```

Comment:
The `.popleft()` call below can fail if the timing of the device is too fast and `commands_epoch` has not yet been populated. Add a check here that loops again until it has.

Something like:

```python
            if not self.commands_epoch:
                log.msg('>>> WAITING FOR COMMANDS_EPOCH <<<')
                return None  # Loop one more time if no commands_epoch?
```